### DISCUSSION TITLE
[Question][Hw1] Having some problem while connecting semantic.ml and domain.ml

### DISCUSSION THREAD

**Discussion by githubqookie** [prosyslab-classroom/cs424-program-reasoning#285] [Q&A]
Name: Kim Seung Gi

I'm trying to connect semantic.ml and domain.ml, but there are some question.

1. I think "transfer" needs to return memory, but I'm not sure how to put some side-effect on "print", "input" while returning memory. The type of function "transfer" changes from "(Llvm.llbasicblock * Memory.t) option -> Llvm.llvalue -> Memory.t -> Memory.t" to (Llvm.llbasicblock * Memory.t) option -> Llvm.llvalue -> unit ->unit" when I put print function (in domain.ml).

2. Is it ok to make new function in domain.ml and semantic.ml? For example, I think we need a function "to_int : t -> int" in domain.ml. Is there an other way to implement the semantics of binary expression without using "to_int" function?

Thank you for reading.

**Comment by goodtaeeun** [comment]
1. The `print` function does not change the memory. However, the memory must continue to be passed on to the next instruction. If you keep these two things in mind, you may figure out how to handle the `print` function in the `transfer` function.
2. First of all, yes, you can add helper functions if they are necessary. However, you can implement the semantics of the binary expressions without functions such as `to_int`. Try the `match` syntax supported by OCaml to deconstruct the type `Int`.

**Comment by githubqookie** [comment]
Thank you!
